add checks to BayesFIT

TDJorgensen opened the pull request:
Hi Ed,

I mostly just cleaned up our code and added logical checks/warnings like I discussed with you last week.  But I also added a `rescale=` argument to request how they are calculated.   `rescale="ppmc"` would rescale the observed-data discrepancies by replicated-data discrepancies (as in Rens' paper), whereas `rescale="devM"` would rescale the observed-data discrepancies by _pD_ to mimic the deviance evaluated at the posterior mean (as in our paper).  Currently, `rescale="ppmc"` returns an error message that it is not yet available.  We will make sure to provide documentation before we request that this function becomes public.

To make them available, I wondered how easy it would be to add `@external$replls` to the fitted blavaan object.  Judging from the source code, probably never going to happen.  

Since the `postpred()` function is only ever called once (only to save PPP, nothing about deviances/`"chisq"` across the posterior), would you be open to me separating that function into smaller components?  I envision the components working together to return the same result, but allowing for more flexibility (some accessible by the user).  So if there is a separate function for generating replicated data from each posterior draw, those could be fed to a separate function to calculate discrepancies (therefore accessible for calculating Rens' BRMSEA).  But it would also allow other user-specified discrepancy functions to be calculated for each posterior draw (e.g., SRMR or model-implied factor correlations, as seen in Levy's 2011 paper). 

I don't plan to have time to start that anytime soon, but please let me know what you think, so I can brainstorm about it.

Thanks,
Terry

ecmerkle commented:
> To make them available, I wondered how easy it would be to add @external$replls to the fitted blavaan object. Judging from the source code, probably never going to happen.

Do you mean that you want the log-likelihood value at each iteration to be saved? If so, I think it is already stored in @external$samplls. Generally, anything could go in @external because it is a simple list.

I am open to splitting up postpred(). postpred() currently has an argument called "measure", that I set up to potentially obtain other things beyond log-likelihood. But it is incomplete. I will also think about it...

TDJorgensen commented:
To clarify, I was thinking that `@external$replls` would be analogous to `@external$samplls`, but contain information about the posterior-predictive samples instead of the observed data.  But Mauricio made an addition to the list returned by the `postpred()` function, which is more straight-forward to access what we need for Rens' BRMSEA.  I'll add that to this pull request tomorrow.

TDJorgensen commented:
I added @maugavilla's updates to `postpred()`, which allows `BayesFIT()` to use the PPMC method of rescaling the chi-squared (i.e., Rens' proposal).  I also added an argument to request a specific _pD_.  This seems to work without issues:
```
hsm0 <- c(paste0("x", 1:9, " ~~ x", 1:9), paste0("x", 1:9, " ~ 1"))
hsm1 <- '
visual  =~ x1 + x2 + x3
textual =~ x4 + x5 + x6
speed   =~ x7 + x8 + x9
'
## fit short chains for a quick check
fit0 <- bcfa(hsm0, data = HolzingerSwineford1939, cp = "fa",  
             n.chains = 2, burnin = 500, sample = 500)
fit1 <- bcfa(hsm1, data = HolzingerSwineford1939, cp = "fa", 
             n.chains = 2, burnin = 500, sample = 500)
## compare ours (rescaling with pD) to theirs
ours <- BayesFIT(fit1, fit_null = fit0, pD = "waic", rescale = "devM")
theirs <- BayesFIT(fit1, fit_null = fit0, pD = "loo", rescale = "PPMC")
mySummary <- function(x) c(M = mean(x), BCI.90 = quantile(x, probs = c(.05, .95)))
apply(ours, 2, mySummary)
apply(theirs, 2, mySummary) # implies better fit when N < 1000, as Rens' paper showed
```

I also added a (currently ignored) `discFUN=` argument for `postpred()` to accept a custom discrepancy function (or a list of such functions), and I made some FIXME comments throughout in places where I thought it would be somewhat easy to apply the custom function to a fitted (b)lavaan object.  For replicated data, this would currently only work in the case of missing data, when lavaan is called to get the EM moments to calculate those log-likelihoods.  But perhaps we could use Yves' same hack when the `discFUN=` argument is non-`NULL`, so that the custom discrepancy function(s) can be applied to the lavaan object.  I know the `measure=` argument already allows for different `fitMeasures()` output to be saved (e.g., to calculate PPP using RMSEA, SRMR, or CFI instead of chi-squared), but `discFUN=` would allow for other discrepancies (e.g., Levy's proposal to compare model-implied factor correlations between observed and replicated data).  If this sounds reasonable to you, I will find some time to implement it after I work on some semTools issues.
